File: abc332b/abc332b.hs
main :: IO ()
main = unwords <$> map show <$> solve <$> f >>= putStrLn
  where
    f = map read <$> words <$> getLine

solve :: [Int] -> [Int]
solve [k, gg, mm] = foldr id [0, 0] $ replicate k f
  where
    f [g, m]
      | g == gg = [0, m]
      | m == 0 = [g, mm]
      | otherwise = let mn = min (gg - g) m in [g + mn, m - mn]

File: abc332b/abc332b.ml
open Printf
open Scanf

let solve k gg mm =
  let rec loop i g m =
    if i = 0 then (g, m)
    else if g = gg then loop (i - 1) 0 m
    else if m = 0 then loop (i - 1) g mm
    else let mn = min (gg - g) m in loop (i - 1) (g + mn) (m - mn) in
  loop k 0 0

let () =
  scanf "%d %d %d " solve |> fun (g, m) -> printf "%d %d\n" g m
